File: command-control/tile.py
from pixel import Pixel
import json

class Tile:
  # Constructor
  def __init__(self, device_name: str):
    print("Tile created: " + device_name)
    self.device_name: str = device_name
    self._online: bool = False
    self._firmware_version: str = ""
    self._hardware_version: str = ""
    self._pinging: bool = False
    self._uptime: int = 0
    self._sounds: list[str] = []
    self._audio_state: int = 0
    self._audio_looping: bool = False
    self._audio_sound: str = ""
    self._audio_volume: int = 0
    self._brightness: int = 0
    self._pixels: list[Pixel] = []
    self._detected: bool = False

  # Properties
  @property
  def online(self) -> bool:
    return self._online
  
  @online.setter
  def online(self, value: bool) -> None:
    print("Tile " + self.device_name + " is now " + ("online" if value else "offline"))
    self._online = value

  @property
  def firmware_version(self) -> str:
    return self._firmware_version

  @property
  def hardware_version(self) -> str:
    return self._hardware_version

  @property
  def pinging(self) -> bool:
    return self._pinging
  
  @property
  def uptime(self) -> int:
    return self._uptime

  @property
  def sounds(self) -> list[str]:
    return self._sounds
  
  @property
  def audio_state(self) -> int:
    return self._audio_state

  @property
  def audio_looping(self) -> bool:
    return self._audio_looping

  @property
  def audio_sound(self) -> str:
    return self._audio_sound

  @property
  def audio_volume(self) -> int:
    return self._audio_volume

  @property
  def brightness(self) -> int:
    return self._brightness

  @property
  def pixels(self) -> list[Pixel]:
    return self._pixels

  @property
  def detected(self) -> bool:
    return self._detected
  
  # Methods
  def update_state(self, state: str) -> None:
    """Set the state of the tile from a JSON string"""

    print("Tile " + self.device_name + " state updated")
    # Convert JSON string to JSON object
    state_json: dict = json.loads(state)

    # TODO: Check if the JSON format is correct

    # System
    system = state_json["system"]
    self._firmware_version = system["firmware"]
    self._hardware_version = system["hardware"]
    self._pinging = system["ping"]
    self._uptime = system["uptime"]
    self._sounds = system["sounds"]

    # Audio
    audio = state_json["audio"]
    self._audio_state = audio["state"]
    self._audio_looping = audio["looping"]
    self._audio_sound = audio["sound"]
    self._audio_volume = audio["volume"]

    # Light
    light = state_json["light"]
    self._brightness = light["brightness"]
    pixels = light["pixels"]
    for i in range(len(pixels)):
      self._pixels[i].from_dict(pixels[i])

    # Detection
    detection = state_json["detect"]
    self._detected = detection["detected"]

File: server/command/lib/views/tile.dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

import '../models/tile_model.dart';

class TilePage extends StatelessWidget {
  const TilePage({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 2, 
      child: Scaffold(
        appBar: AppBar(
          backgroundColor: Theme.of(context).colorScheme.primary,
          foregroundColor: Theme.of(context).colorScheme.onPrimary,
          leading: IconButton(
            icon: const Icon(Icons.arrow_back),
            tooltip: 'Home',
            onPressed: () => context.go("/"), // Go back to home screen
          ),
          bottom: const TabBar(
              tabs: [
                Tab(text: 'Command'),
                Tab(text: 'State'),
              ],
              labelColor: Colors.white,
              unselectedLabelColor: Colors.white60,
            ),
          title: Text(tile.deviceName),
          centerTitle: true,
        ),
        body: TabBarView(
          children: [
            TileCommand(tile: tile),
            TileState(tile: tile),
          ],
        ),
      ),
    );
  }
}

class TileCommand extends StatelessWidget {
  const TileCommand({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return 
    Column(
      children: [
        Expanded(child: SystemCommand(tile: tile)),
        const Divider(
              indent: 20,
              endIndent: 20,
            ),
        Expanded(child: AudioCommand(tile: tile)),
        const Divider(
              indent: 20,
              endIndent: 20,
            ),
        Expanded(child: LightCommand(tile: tile)),
        // add a send command button
        FloatingActionButton(
          child: const Icon(Icons.send),
          onPressed: () {
            // Add your onPressed code here!
          },
        )
      ],
    );
  }
}

class SystemCommand extends StatelessWidget {
  const SystemCommand({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const Center(child: Text('System Command')),
        Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text('Pinging: '),
            Switch(
              value: tile.pinging,
              onChanged: (value) {
                // Add your onChanged code here!
              },
              activeTrackColor: Colors.lightGreenAccent,
              activeColor: Colors.green,
            ),
          ],
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text('Reboot: '),
            Switch(
              value: tile.pinging, // TODO: Add reboot to model
              onChanged: (value) {
                // Add your onChanged code here!
              },
              activeTrackColor: Colors.lightGreenAccent,
              activeColor: Colors.green,
            ),
          ],
        )
      ],
    );
  }
}

class AudioCommand extends StatelessWidget {
  const AudioCommand({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const Center(child: Text('Audio Command')),
        Center(child: Text('Mode: None')),
        Center(child: Text('Looping: ${tile.audioLooping}')),
        Center(child: Text('Sound: ${tile.audioSound}')),
        Center(child: Text('Volume: ${tile.audioVolume}')),
      ],
    );
  }
}

class LightCommand extends StatelessWidget {
  const LightCommand({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const Center(child: Text('Light Command')),
        Center(child: Text('Brightness: ${tile.brightness}')),
        const Center(child: Text('Pixels:')),
        for (var i = 0; i < tile.pixels.length; i++)
          Center(child: Text(tile.pixels[i].toJson().toString())),
      ],
    );
  }
}

class TileState extends StatelessWidget {
  const TileState({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return 
    
    Column(
      children: [
        Expanded(child: SystemState(tile: tile)),
        const Divider(
              indent: 20,
              endIndent: 20,
            ),
        Expanded(child: AudioState(tile: tile)),
        const Divider(
              indent: 20,
              endIndent: 20,
            ),
        Expanded(child: LightState(tile: tile)),
        const Divider(
              indent: 20,
              endIndent: 20,
            ),
        Expanded(child: PresenceState(tile: tile)),
      ],
    );
  }
}

class SystemState extends StatelessWidget {
  const SystemState({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const Center(child: Text('System State')),
        Center(child: Text('Online: ${tile.online}')),
        Center(child: Text('Firmware Version: ${tile.firmwareVersion}')),
        Center(child: Text('Hardware Version: ${tile.hardwareVersion}')),
        Center(child: Text('Pinging: ${tile.pinging}')),
        Center(child: Text('Uptime: ${tile.uptime}')),
        Center(child: Text('Sounds: ${tile.sounds}')),
      ],
    );
  }
}

class AudioState extends StatelessWidget {
  const AudioState({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const Center(child: Text('Audio State')),
        Center(child: Text('State: ${tile.audioState}')),
        Center(child: Text('Looping: ${tile.audioLooping}')),
        Center(child: Text('Sound: ${tile.audioSound}')),
        Center(child: Text('Volume: ${tile.audioVolume}')),
      ],
    );
  }
}

class LightState extends StatelessWidget {
  const LightState({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const Center(child: Text('Light State')),
        Center(child: Text('Brightness: ${tile.brightness}')),
        const Center(child: Text('Pixels:')),
        for (var i = 0; i < tile.pixels.length; i++)
          Center(child: Text(tile.pixels[i].toJson().toString())),
      ],
    );
  }
}

class PresenceState extends StatelessWidget {
  const PresenceState({super.key, required this.tile});

  final Tile tile;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        const Center(child: Text('Presence State')),
        Center(child: Text('Detected: ${tile.detected}')),
      ],
    );
  }
}